shell.nim

```diff
@@ -220,31 +251,65 @@ proc asgnShell*(cmd: string): tuple[output: string, exitCode: int] =
         # outstream died on us?
         doAssert outStream.isNil
         break
+
+    if not outStream.atEnd():
+      if dokOutput in debugConfig:
+        let rem = outStream.readAll()
+        res &= rem
+        for line in rem.split("\n"):
+          echo "shell> ", line
+      else:
+        res &= outStream.readAll()
+
     let exitCode = pid.peekExitCode
+
+    # Zero exit code does not guarantee that there will be nothing in
+    # stderr.
+
+    let err = pid.errorStream
+    let errorText = err.readAll()
+    err.close()
+
     if exitCode != 0:
-      # add error stream to output
-      let err = pid.errorStream
-      res.add err.readAll()
-      err.close()
+      if dokRuntime in debugConfig:
+        echo "Error when executing: ", cmd
+
+      if dokError in debugConfig:
+        for line in errorText.split("\n"):
+          echo "err> ", line
+
     pid.close()
-    result = (output: res, exitCode: exitCode)
+    result = (output: res, error: errorText, exitCode: exitCode)
   else:
     # prepend the NimScript called command by current directory
     let nscmd = &"cd {getCurrentDir()} && " & cmd
-    result = gorgeEx(nscmd, "", "")
-  result[0] = result[0].strip(chars = {'\n'})
-
-proc execShell*(cmd: string): tuple[output: string, exitCode: int] =
+    let (res, code) = gorgeEx(nscmd, "", "")
+    result.output = res
+    result.exitCode = code
+  result.output = result.output.strip(chars = {'\n'})
+  result.error = result.error.strip(chars = {'\n'})
+
+proc execShell*(
+  cmd: string,
+  debugConfig: set[DebugOutputKind] = defaultDebugConfig
+              ): tuple[output, error: string, exitCode: int] =
   ## wrapper around `asgnShell`, which calls the commands and handles
   ## return values.
-  echo "shellCmd: ", cmd
-  result = asgnShell(cmd)
-  when defined(NimScript):
-    # output of child process is already echoed on the fly for non NimScript
-    # usage
-    if result[0].len > 0:
-      for line in splitLines(result[0]):
-        echo "shell> ", line
+
+  # IDEA split command by space and layout it line-by-line if it is
+  # too long to fit on one line. This will increase readability of
+  # debug output
```

Comment:
I guess it would make sense to split the input command wherever we (likely) introduced manual concatenation via `&&` of several commands. That way it directly corresponds to the actual user input to the macro (unless one manually uses `&&` of course).

Reply:
I don't think it is necessary to cram everything I want to do in one PR. If necessary I will open another one.

About splitting by `&&` vs spaces: it all comes down to use cases. Right now I mostly use with commands that require to pass a lot of different and long arguments (`latexmk` for example) together with absolute path for a file. Even if there is a single command it might take several lines in terminal. 

So it all comes down to use cases and it will requires some testing to get it look nice/